apisix/admin/stream_routes.lua

```diff
@@ -25,23 +27,69 @@ local _M = {
     need_v3_filter = true,
 }
 
+local function check_router_refer(items, id)
+    local refer_list = {}
+    local referkey = "/stream_routes_refer/" .. id
+    local _, err = core.etcd.delete(referkey)
+    core.log.warn(err)
+    core.log.warn(items)
+    for _, item in config_util.iterate_values(items) do
+        if item.value == nil then
+            goto CONTINUE
+        end
+        local r_id = string.gsub(item["key"], "/", "_")
+        local route = item.value
+        if route.protocol and route.protocol.superior_id then
+            local data
+            local setkey = "/stream_routes_refer/" .. route.protocol.superior_id
+            local res, err = core.etcd.get(setkey, false)
+            if res then
+                if res.body.node == nil then
+                    data = core.json.decode("{}")
+                    data[r_id] = 1
+                else
+                    data = res.body.node.value
+                    data[r_id] = 1
+                end
+            else
+                core.log.error("In function check_router_refer  error: ", err)
+            end
+            local setres, err = core.etcd.set(setkey, data)
```

Comment:
We should avoid writing this data to etcd. You set here, I don't see where we delete key. so `stream_routes_refer` will grow forever?

Reply:
The keys  stream_routes_refer/super_ids  will be update   rather than grow .  
ex：

/apisix/stream_routes_refer/12
{"_apisix_stream_routes_22":1,"_apisix_stream_routes_23":1}  

it meas the stream_route 22 and 23  have   protocol.superior_id  12; if the stream_routes rule  no change ,the value no change.

Reply:
so at what point do we delete `/apisix/stream_routes_refer/12`?

Anyway, I don't think it's a good solution to write this relationship to etcd. 

Can we refer to the logic for removing the upstream?https://github.com/apache/apisix/blob/a62d33f2f1fcbd62fdd25ea2295b0d6c4675bd4c/apisix/admin/upstreams.lua#L146-L156

Reply:
Thanks for your suggestion I changed the implementation.